From 96a06f4bfb87e8af1a8446f75ef7795c47557c91 Mon Sep 17 00:00:00 2001
From: Emmanuel Blot <emmanuel.blot@free.fr>
Date: Tue, 21 Jul 2026 14:33:37 +0200
Subject: [PATCH 698/705] tests/functional/arm: minerva: test INA238

Extend the Minerva CMM boot test to validate an INA238 power monitor on
Fan Controller Board 1 (i2c2 PCA9548 channel 1, address 0x40): drive the
bus voltage and die temperature through QOM and check each is reported
back via the guest's kernel hwmon (ina238) interface.

Signed-off-by: Emmanuel Blot <emmanuel.blot@free.fr>
---
 tests/functional/arm/test_aspeed_minerva.py | 23 +++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/tests/functional/arm/test_aspeed_minerva.py b/tests/functional/arm/test_aspeed_minerva.py
index 738e26bae4..b36aad1ed1 100644
--- a/tests/functional/arm/test_aspeed_minerva.py
+++ b/tests/functional/arm/test_aspeed_minerva.py
@@ -24,6 +24,14 @@ class MinervaMachine(AspeedTest):
     # bus, hence child[1] (child[0] is the controller's own slave).
     INA230_QOM_PATH = "/machine/soc/i2c/bus[0]/aspeed.i2c.bus.0/child[1]"
     INA230_HWMON = "/sys/bus/i2c/devices/0-0040/hwmon/hwmon*"
+
+    # The FCB INA238 power monitors sit behind the i2c2 PCA9548 mux. FCB 1 is
+    # mux channel 1 (aliased i2c16 in the device tree); on that channel bus the
+    # eeprom@50 is child[0] and the 0x40 monitor is the next slave, child[1].
+    # In the guest it therefore enumerates as bus 16, device 16-0040.
+    INA238_QOM_PATH = ("/machine/soc/i2c/bus[2]/aspeed.i2c.bus.2"
+                       "/child[1]/i2c.1/child[1]")
+    INA238_HWMON = "/sys/bus/i2c/devices/16-0040/hwmon/hwmon*"
     PROMPT = "root@minerva:~#"
 
     def test_arm_ast2600_minerva_openbmc(self):
@@ -51,6 +59,21 @@ def test_arm_ast2600_minerva_openbmc(self):
             self.wait_hwmon_value(self.INA230_HWMON, "in1_input",
                                   bus_uv // 1000)
 
+        self.assertIn(b"ina238", self.read_hwmon(self.INA238_HWMON, "name"))
+
+        # in1_input reports the bus voltage in mV (3.125 mV/LSB) and temp1_input
+        # the die temperature in millidegrees C (125 m-degC/LSB). Drive both
+        # through QOM and check they read back through the kernel
+        # hwmon interface.
+        for bus_uv, temp_mc in ((12000000, 30000), (3300000, 55000)):
+            self.vm.cmd("qom-set", path=self.INA238_QOM_PATH,
+                        property="bus-voltage", value=bus_uv)
+            self.vm.cmd("qom-set", path=self.INA238_QOM_PATH,
+                        property="die-temperature", value=temp_mc)
+            self.wait_hwmon_value(self.INA238_HWMON, "in1_input",
+                                  bus_uv // 1000)
+            self.wait_hwmon_value(self.INA238_HWMON, "temp1_input", temp_mc)
+
 
 if __name__ == '__main__':
     AspeedTest.main()
-- 
2.50.1

